### PR TITLE
DOC: Adds Iñigo's week 3 and 4 blogpost

### PR DESCRIPTION
**Pull request by itellaetxe**
Adds @itellaetxe blogpost for Week 3.

### CODE REVIEW

**Comment by itellaetxe** [comment]
@skoudoro I still cannot request review from anyone in this repo. In the DIPY repo, I can, but not here.
Also, I don't know why the previous commits are still showing, I squashed them...

File: posts/2024/2024_06_14_Inigo_week_3.rst [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+I also worked on trying to monitor the training experiments using TensorBoard, but I did not succeed because it was a last minute idea and I did not have time to implement it properly.
+
+.. image:: /_static/images/fibercup_better_results.png
+    :alt: Preliminary results of the AutoEncoder training for a subset of plausible fibers of the FiberCup dataset, approaching better replication compared to the PyTorch model.
+    :width: 600
+
+What is coming up next week
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+My mentors and I agreed on trying to transfer the weights of the pre-trained PyTorch model to my Keras implementation, because it may take less time than actually training the model. Thus, the strategy we devised for this to work is the following:
+1. Implement dataset loading using HDF5 files, as the original model uses them, and the TractoInferno dataset is contained in such files (it is approximately 75 GB)/
```
Comment (robinroy03) [inline posts/2024/2024_06_14_Inigo_week_3.rst:25]: `/` looks like a typo for the punctuation mark `.`

**Review by robinroy03 — APPROVED** [review]
LGTM. Good work 👍. Please see the minor typo below but that's everything.

File: posts/2024/2024_06_14_Inigo_week_3.rst (hunk 1 above)
Comment (itellaetxe) [reply posts/2024/2024_06_14_Inigo_week_3.rst:25]: True, TY for your review :)!

**Review by jhlegarreta — APPROVED** [review]

**Comment by deka27** [comment]
LGTM. Cant review tho.

**Comment by itellaetxe** [comment]
This PR had not been merged and I did not want to create another branch in my fork, so I just pushed the new blog here (week 4). I also updated the PR title.

@WassCodeur, @deka27, I think you cannot review this for whatever reason, but could you comment something on the PR? (I already saw your comment on week 3 @deka27, ty)

@skoudoro, whenever you can, could you check the review request problem?

thanks!

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 2]
```diff
@@ -0,0 +1,48 @@
+Week 4 into GSoC 2024: Weight transfer experiments, hardships, and results!
```
Comment (WassCodeur) [inline posts/2024/2024_06_21_Inigo_week_4.rst:1]: Hi @itellaetxe !

Great work,

Comment (WassCodeur) [reply posts/2024/2024_06_21_Inigo_week_4.rst:1]: I've added a few comments below

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+===========================================================================
+
+.. post:: June 21 2024
+    :author: Iñigo Tellaetxe
+    :tags: google
+    :category: gsoc
+
+
+What I did this week
+~~~~~~~~~~~~~~~~~~~~
```
Comment (WassCodeur) [inline posts/2024/2024_06_21_Inigo_week_4.rst:11]: It's just that it's a bit tight, so if you could put spaces, it would be better for reading.

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+The errors in the Decoder block are higher than in the Encoder but we assumed that a MAE of around 1e-3 is acceptable.
+On the other hand I started implementing the dataset loading using HDF5 files, but I set that aside because it is not priority.
+Finally, my mentor `Jon Haitz <https://github.com/jhlegarreta>`_ kindly provided me with the weights of the PyTorch AE he trained on the FiberCup dataset, and he suggested an experiment consisting of encoding the FiberCup tractogram with my Keras model, and Decoding it with the PyTorch model to see if the Encoder works properly.
+This was indeed the case, as the PyTorch model effectively reconstructed the tractogram, but unfortunately the Keras encoder was not capable of giving the same result. Naturally, this suggests that the Keras Decoder implementation is still not similar enough to the PyTorch one, so there is still room
+for improvement. Despite not being successful, this experiment was very enlightening, and it gave me a lot of insight into the differences between the two implementations.
```
Comment (WassCodeur) [inline posts/2024/2024_06_21_Inigo_week_4.rst:28]: Cool, it's clear even not being too familiar I get it. Thank you for sharing

File: posts/2024/2024_06_14_Inigo_week_3.rst [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+===================================================================================
+
+.. post:: June 14 2024
+    :author: Iñigo Tellaetxe
+    :tags: google
+    :category: gsoc
+
+
+What I did this week
+~~~~~~~~~~~~~~~~~~~~
```
Comment (WassCodeur) [inline posts/2024/2024_06_14_Inigo_week_3.rst:11]: LGTM, apart from that if you could leave a blank line after the titles, like here, it would be better for reading.

**Review by WassCodeur — APPROVED** [review]

File: posts/2024/2024_06_21_Inigo_week_4.rst (hunk 2 above)
Comment (itellaetxe) [reply posts/2024/2024_06_21_Inigo_week_4.rst:1]: TY for the review Wachiou!

**Review by WassCodeur — APPROVED** [review]

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+This was partially good news, because most of the Encoder was behaving as desired, but, the most challenging part was adapting the flattening and reshaping operations happening in the encoder and the decoder, respectively.
+As the Keras 1D convolutional output dimensions do not follow the same ordering as in PyTorch, (*[n, m, channels]* vs *[n, channels, m]*), the flattening behavior of the models was different, and thus, the fully connected layer of the Encoder (named ``fc1``) was receiving different inputs.
+To solve this, I first reshaped the output of the Keras 1D convolutional layer to match the PyTorch *channels first* convention, and then applied the flattening.
+This effectively resulted in a within-reasonable-error (MAE = 1e-6) output of the Encoder block. Problem solved!
+The Decoder block was a bit more challenging, because the PyTorch implementation was using linear interpolation in its ``torch.nn.Upsample`` layers. For this, I had to implement a custom layer in Keras that would perform the same operation,
```
Comment (jhlegarreta) [inline posts/2024/2024_06_21_Inigo_week_4.rst:24]: Unnecessary line break.

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+as linear interpolation is unavailable in the ``tf.keras.layers.UpSampling1D`` layers. I made this workaround using the ``tf.image.resize`` function, tricking the function into taking a modified 1D tensor to be a pseudo 2D tensor.
+The errors in the Decoder block are higher than in the Encoder but we assumed that a MAE of around 1e-3 is acceptable.
+On the other hand I started implementing the dataset loading using HDF5 files, but I set that aside because it is not priority.
+Finally, my mentor `Jon Haitz <https://github.com/jhlegarreta>`_ kindly provided me with the weights of the PyTorch AE he trained on the FiberCup dataset, and he suggested an experiment consisting of encoding the FiberCup tractogram with my Keras model, and Decoding it with the PyTorch model to see if the Encoder works properly.
+This was indeed the case, as the PyTorch model effectively reconstructed the tractogram, but unfortunately the Keras encoder was not capable of giving the same result. Naturally, this suggests that the Keras Decoder implementation is still not similar enough to the PyTorch one, so there is still room
```
Comment (jhlegarreta) [inline posts/2024/2024_06_21_Inigo_week_4.rst:29]: Unnecessary line break.

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+.. image:: /_static/images/fibercup_replicated.png
+    :alt: Left: Source data to encode/decode. Middle: Keras model reconstruction. Right: PyTorch model reconstruction.
+    :width: 600
+
+
+What is coming up next week
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Next week we will start working on a conditional version of the AutoEncoder, which should give us the ability to generate tractograms conditioned on a specific scalar input. This will be a very interesting feature to have because we can get tractograms with properties of interest. Well, this is the main goal of this project.
+The sampling strategy for the tracts does not concern me for now because the code is already available in the `tractolearn <https://github.com/scil-vital/tractolearn>`_ repository, so we can postpone it for now.
```
Comment (jhlegarreta) [inline posts/2024/2024_06_21_Inigo_week_4.rst:43]: "The sampling strategy for the tracts does not concern me for now because the code is" : "It is decided to focus on developping a conditional version of the autoencoder over adding the latent space sampling because the code for the latter is"

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+After doing this, I set a common input for both networks, and I compared their outputs. As expected, they were not the same, and they were not even close enough.
+Thus, matching the behavior of the PyTorch model with the Keras implementation became my objective. To achieve so, I run a common input through all the layers of both models sequentially, and systematically compared the outputs of each layer.
+In the Encoder block, I found all the outputs to be within a reasonable range of each other (MAE = 1e-6), except for the last two operations, which flatten the output of the 1D convolutional layers and then feed it to a fully connected layer.
+This was partially good news, because most of the Encoder was behaving as desired, but, the most challenging part was adapting the flattening and reshaping operations happening in the encoder and the decoder, respectively.
+As the Keras 1D convolutional output dimensions do not follow the same ordering as in PyTorch, (*[n, m, channels]* vs *[n, channels, m]*), the flattening behavior of the models was different, and thus, the fully connected layer of the Encoder (named ``fc1``) was receiving different inputs.
```
Comment (jhlegarreta) [inline posts/2024/2024_06_21_Inigo_week_4.rst:21]: the flattening behavior of the models was different (the elements followed a different sorting when being concatenated into a 1D array)

If the above is accurate as to what was happening.

**Review by jhlegarreta — COMMENTED** [review]
Thanks for this work.

Contents are well-explained. Thanks.

Two comments about the form:
- Not sure how this looks like when it is deployed since I am unable to find the corresponding URL, but the text looks busy: I would group related sentences in real paragraphs and add white lines between paragraphs, otherwise it is hard to follow.
- Although it is not that important here, remember that adding a commit message body is helpful. Let me know if you are unsure how to do this.

**Comment by skoudoro** [comment]
> @skoudoro, whenever you can, could you check the review request problem?

it should be ok now, Can you confirm ?

> Not sure how this looks like when it is deployed since I am unable to find the corresponding URL, but the text looks busy

I agree with @jhlegarreta. Also, one CI's seems to stop to work, that's why we do not have the URL. I need to check what's going on.

Also, this PR is not compiling.  It would be great if you could try to compile locally @itellaetxe before requesting review. You will get early on all the errors and the process will be faster.

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 10]
```diff
@@ -0,0 +1,53 @@
+Week 4 into GSoC 2024: Weight transfer experiments, hardships, and results!
```
Comment (robinroy03) [inline posts/2024/2024_06_21_Inigo_week_4.rst:1]: We'll need a paragraph break here. (Image from `blog.html`)

image

**Review by robinroy03 — CHANGES_REQUESTED** [review]
Hi, please see the comments below.

**Comment by itellaetxe** [comment]
@skoudoro, I can now request reviewers, I added you, @pjsjongsung, and @deka27. Thank you.

@jhlegarreta, you are right and I remember you had already told me about this, but I did not look into it, sorry. I will correct the blog and add the commit body message. (For reference about the commit body message, I looked at [this answer in stackoverflow].(https://stackoverflow.com/a/36427485)

@robinroy03 thanks for the feedback, correcting it now.

File: posts/2024/2024_06_21_Inigo_week_4.rst (hunk 8 above)
Comment (itellaetxe) [reply posts/2024/2024_06_21_Inigo_week_4.rst:43]: Sounds way better, thank you

File: posts/2024/2024_06_21_Inigo_week_4.rst [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+This was indeed the case, as the PyTorch model effectively reconstructed the tractogram, but unfortunately the Keras encoder was not capable of giving the same result. Naturally, this suggests that the Keras Decoder implementation is still not similar enough to the PyTorch one, so there is still room
+for improvement. Despite not being successful, this experiment was very enlightening, and it gave me a lot of insight into the differences between the two implementations.
+In a last effort to get to replicate the PyTorch model results, I went on to train the my Keras architecture on the FiberCup dataset with the same parameters as my mentor used in his `GESTA <https://doi.org/10.1016/j.media.2023.102761>`_ paper to see if the results I get are similar to the ones he got.
+Well, this resulted in amazing results, as you can check visually in the figure below. Note that none of the models were able to capture the depth dimension of the streamlines, but this is not concerning. It can be solved reducing the latent dimension size to 16 (it is 32 now).
```
[the remark this answers is not in the document]
Comment (itellaetxe) [reply posts/2024/2024_06_21_Inigo_week_4.rst:32]: I remember I made an experiment with the latent dimension size set to 16, and the depth was captured. While I agree that it is indeed counterintuitive, I got that result. However, since I did not prove thoroughly that the problem is solved like this, I will just stick to saying, "it can be solved by changing the latent dimension size" to keep it more general.

File: posts/2024/2024_06_21_Inigo_week_4.rst (hunk 9 above)
Comment (itellaetxe) [reply posts/2024/2024_06_21_Inigo_week_4.rst:21]: Yes, correct. Added the comment in parenthesis, it clarifies it further. TY

File: posts/2024/2024_06_21_Inigo_week_4.rst (hunk 10 above)
Comment (itellaetxe) [reply posts/2024/2024_06_21_Inigo_week_4.rst:1]: Thanks for the comment @robinroy03, but I am not sure how to add it. Could you give me a hint, please?

File: posts/2024/2024_06_21_Inigo_week_4.rst (hunk 6 above)
Comment (itellaetxe) [reply posts/2024/2024_06_21_Inigo_week_4.rst:24]: Removed

File: posts/2024/2024_06_21_Inigo_week_4.rst (hunk 7 above)
Comment (itellaetxe) [reply posts/2024/2024_06_21_Inigo_week_4.rst:29]: Removed

File: posts/2024/2024_06_21_Inigo_week_4.rst (hunk 10 above)
Comment (robinroy03) [reply posts/2024/2024_06_21_Inigo_week_4.rst:1]: @itellaetxe your new commit fixes it. The first paragraph will be highlighted on the blog page.

**Review by skoudoro — APPROVED** [review]
Thanks all for all the review. 

week 4 is still a bit dense and needs more empty lines.

However, I will go ahead and merge this PR.

Note: the URL was not appearing because the blog compilation was failing. At the end, everything works almost as expected but when it fails, the CI's should fails. I will try to look in to it this week

thanks @itellaetxe
